### DISCUSSION TITLE
Can't create Star with specific epoch.

### DISCUSSION THREAD

**Discussion by AstroKeith** [skyfielders/python-skyfield#1023] [Q&A] [answered]
My goal is to calculate the current AltAz & rates from a measured RA & Dec (ie at JNow), so I can control a AltAz telescope. But the calculated Az & Alt are out by a few arc minutes. So I suspected precession. I put some prints in my code and found that the epoch I am specifying when creating the Star is ignored (or so it seems!).

Here is my test code ....
def get_altaz(self,ra,dec):
        t = self.ts.now()
        scope = Star(ra_hours=ra,dec_degrees=dec,epoch=t)
        a = self.get_location().at(t).observe(scope).apparent()
        alt, az, distance, alt_rate, az_rate, range_rate = (a.frame_latlon_and_rates(self.site))
        ra,dec,d = a.radec()
        print('radec',ra,dec) # used to check whether input RA & Dec are 
        print('AltAz',alt,az)

self.location & self.site are ..

self.location = self.coordinates.get_earth() + wgs84.latlon(self.lat, self.long)
       self.site = wgs84.latlon(self.lat,self.long)

I can change epoch=t, to say epoch=self.ts.tt(2000.0) and it has no affect on any output figures

As a test I compared the results for two versions of the line 'scope = Star(....

scope = Star(ra_hours=(5,55,09.4),dec_degrees=(7,24,19.8)) # Betelgeuse at J2000
with
scope = Star(ra_hours=(5,56,31.08),dec_degrees=(7,24,43,3),epoch=t) # today

I would have expected the 'Star' created by each to be the same, as the later one is just Betelgeuse precessed.
But only the first one gives the correct AltAz answer.

My problem is that I need sometimes to be able to input current RA & Dec

Any help appreciated!

**Comment by brandon-rhodes** [comment, marked as answer]
Good question! Skyfield is, alas, not entirely consistent in its use of `epoch` (following many earlier astronomy computer libraries which did the same thing). "Epoch" means "a moment in time", which in this case means "the moment the star was at this RA and dec." So you will only see any effect from `epoch` if you specify a velocity for the star — since, otherwise, it's assumed to be at that RA and dec forever.

But you are wanting its other sense — "epoch" meaning "the moment I'm choosing to make a coordinate system out of, using the Earth's equator, poles, and equinox as of that moment." So B1950 coordinates can be said to have an "epoch" of 1950, and J2000 coordinates an "epoch" of the year 2000, and so forth.

But it's confusing to use "epoch" for both (a) the moment we measure the position of a star that's in motion across the sky, and (b) the moment in the Earth's long history of precession that we use to fix our RA/dec coordinate system. Better authorities today only use "epoch" for the first meaning. For the second, they instead say "equinox", like "equinox 1950.0", or "equinox of the year 2000".

My guess is that you should try out:

https://rhodesmill.org/skyfield/coordinates.html#turning-coordinates-into-a-position

— Skyfield's `from_radec()` method, that lets you specify not only an RA and dec, but the equinox date for the Earth RA/dec coordinate system you want to use. Alas! Because I wasn't careful enough back when I added it to Skyfield, instead of naming the parameter `equinox=`, I continued on with the old confusion and named it "epoch":

```
    @classmethod
    def from_radec(cls, ra_hours, dec_degrees,
                   distance_au=_GIGAPARSEC_AU, epoch=None):
```

But it should hopefully let you create an x,y,z position whose RA and dec shift with precession, instead of being fixed like the coordinates in star catalogs. Try it out, and let me know if it lets you make any headway!

**Comment by AstroKeith** [comment]
Thanks you so much for the quick response.
That works!

Here is my implementation, I assuming I still have to create a 'Star', but at least I'm building at J2000.
```
def get_rate(self,raNow,decNow):
        t = self.ts.now()   
        position = Apparent.from_radec(ra_hours=raNow, dec_degrees=decNow, epoch=t)
        ra,dec,d = position.radec() # this is at J2000
        scope = Star(ra_hours=ra.hours, dec_degrees=dec.degrees)
        a = self.get_location().at(t).observe(scope).apparent()
        alt, az, distance, alt_rate, az_rate, range_rate = (a.frame_latlon_and_rates(self.site))
        return (alt, az, alt_rate, az_rate)
```
